File: Laborator/2020-2021/lab02 - Greedy/skel/cpp/task-1/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <iomanip>
using namespace std;

struct Object {
    int weight;
    int price;

    Object(int _weight, int _price) : weight(_weight), price(_price) {}
};

class Task {
 public:
    void solve() {
        read_input();
        print_output(get_result());
    }

    static bool cmp(Object obj1, Object obj2) {
        return ((double)obj1.price / obj1.weight) > ((double)obj2.price / obj2.weight);
    }

 private:
    int n, w;
    vector<Object> objs;

    void read_input() {
        ifstream fin("in");
        fin >> n >> w;
        for (int i = 0, weight, price; i < n; i++) {
            fin >> weight >> price;
            objs.push_back(Object(weight, price));
        }
        fin.close();
    }

    double get_result() {
        // TODO: Aflati profitul maxim care se poate obtine cu obiectele date.
        
        // sortare descrescatoare dupa pi/wi
        sort(objs.begin(), objs.end(), cmp);

        double profit = 0;
        vector<Object>::iterator it;
        for(it = objs.begin(); it != objs.end() && w; ++it) {
            if ((*it).weight <= w) { // iau tot obiectul
                profit += (*it).price;
                w -= (*it).weight;
            } else { // iau o fractiune din obiect
                profit += (double)(*it).price / (*it).weight * w;
                w = 0;
            }
        }

        return profit;
    }

    void print_output(double result) {
        ofstream fout("out");
        fout << setprecision(4) << fixed << result;
        fout.close();
    }
};

// Please always keep this simple main function!
int main() {
    // Allocate a Task object on heap in order to be able to
    // declare huge static-allocated data structures inside the class.
    Task *task = new Task();
    task->solve();
    delete task;
    return 0;
}

File: Laborator/2020-2021/lab02 - Greedy/skel/java/task-1/src/Main.java
import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Arrays;
import java.util.Comparator;
import java.util.Scanner;

public class Main {
    static class Obj {
        public int weight;
        public int price;

        public Obj() {
            weight = 0;
            price = 0;
        }
    };

    static class Task {
        public final static String INPUT_FILE = "in";
        public final static String OUTPUT_FILE = "out";

        int n, w;
        Obj[] objs;

        public void solve() {
            readInput();
            writeOutput(getResult());
        }

        private void readInput() {
            try {
                Scanner sc = new Scanner(new File(INPUT_FILE));
                n = sc.nextInt();
                w = sc.nextInt();
                objs = new Obj[n];
                for (int i = 0; i < n; i++) {
                    objs[i] = new Obj();
                    objs[i].weight = sc.nextInt();
                    objs[i].price = sc.nextInt();
                }
                sc.close();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

        private void writeOutput(double result) {
            try {
                PrintWriter pw = new PrintWriter(new File(OUTPUT_FILE));
                pw.printf("%.4f\n", result);
                pw.close();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

        private double getResult() {
            // TODO: Aflati profitul maxim care se poate obtine cu obiectele date.

            // sortare descrescatoare dupa price/weight
            Arrays.sort(objs, new Comparator<Obj>() {
                @Override
                public int compare(Obj obj1, Obj obj2) {
                    return ((double)obj1.price / obj1.weight < (double)obj2.price / obj2.weight) ? 1 : -1;
                }
            });

            double profit = 0.0;
            for (Obj obj : objs) {
                if(obj.weight <= w) {
                    profit += obj.price;
                    w -= obj.weight;
                } else {
                    profit += (double)obj.price / obj.weight * w;
                    w = 0;
                    break;
                }
            }
            return profit;
        }
    }

    public static void main(String[] args) {
        new Task().solve();
    }
}
